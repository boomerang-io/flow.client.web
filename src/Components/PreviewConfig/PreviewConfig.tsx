import React from "react";
import {
  DynamicFormik,
  ModalForm,
  ComposedModal,
  Button,
  ModalBody,
  TooltipHover,
} from "@boomerang-io/carbon-addons-boomerang-react";
import TextEditorModal from "Components/TextEditorModal";
import { TEXT_AREA_TYPES } from "Constants/formInputTypes";
import { ComposedModalChildProps, DataDrivenInput, ModalTriggerProps } from "Types";
import { View16 } from "@carbon/icons-react";

const modalHeadertext =
  "This is a preview of what the user sees when editing this Task. The user can also give this task a custom name for their Workflow, and can adjust its connected tasks. You can type in these fields to test any validation requirements.";


const TextEditorInput: React.FC<any> = (props) => {
  return (
    <div key={props.id} style={{ position: "relative", cursor: "pointer", paddingBottom: "1rem" }}>
      <TextEditorModal {...props} {...props.item} />
    </div>
  );
};

const textAreaProps = ({ input, formikProps }: { input: DataDrivenInput, formikProps: any}) => {
  const { values, setFieldValue } = formikProps;
  const { key, type, ...rest } = input;
  //@ts-ignore
  const itemConfig = TEXT_AREA_TYPES[type];
  return {
    autoSuggestions: [],
    formikSetFieldValue: (value: any) => setFieldValue(key, value),
    initialValue: values[key],
    item: input,
    ...itemConfig,
    ...rest,
  };
};

interface PreviewConfigFormProps {
  closeModal: () => void;
  templateConfig: any;
}

const PreviewConfigForm: React.FC<PreviewConfigFormProps> = ({ templateConfig }) => {
  return (
    <DynamicFormik
      allowCustomPropertySyntax
      validateOnMount
      inputs={templateConfig}
      dataDrivenInputProps={{
        TextEditor: TextEditorInput,
      }}
      textEditorProps={textAreaProps}
      toggleProps={() => ({
        orientation: "vertical",
      })}
    >
      {({ inputs }: any) => (
        <ModalForm noValidate>
          <ModalBody>{inputs}</ModalBody>
        </ModalForm>
      )}
    </DynamicFormik>
  );
}

interface PreivewConfigProps {
  templateConfig: any;
  taskTemplateName: string;
}

const PreviewConfig: React.FC<PreivewConfigProps> = ({ templateConfig, taskTemplateName }) => {
  return (
    <ComposedModal
      composedModalProps={{ shouldCloseOnOverlayClick: true }}
      modalHeaderProps={{
        title: `[Preview] ${taskTemplateName}`,
        subtitle: modalHeadertext,
      }}
      modalTrigger={({ openModal }: ModalTriggerProps) => (
        <TooltipHover direction="top" tooltipText={"Preview what the user sees when they view this task"}>
          <Button
            iconDescription="Preview task"
            renderIcon={View16}
            onClick={openModal}
            size="field"
            kind="ghost"
            style={{ width: "6.25rem" }}
          >
            Preview
          </Button>
        </TooltipHover>
      )}
    >
      {({ closeModal }: ComposedModalChildProps) => <PreviewConfigForm templateConfig={templateConfig} closeModal={closeModal} />}
    </ComposedModal>
  );
}
export default PreviewConfig;
