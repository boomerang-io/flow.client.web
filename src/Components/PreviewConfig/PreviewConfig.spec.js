import React from "react";
import { fireEvent } from "@testing-library/react";
import PreviewConfig from "./index";

const props = {
  taskTemplateName: "Test template",
  templateConfig: [
    {
      placeholder: "",
      readOnly: false,
      description: "",
      key: "path",
      label: "File Path",
      type: "text",
    },
    {
      placeholder: "",
      readOnly: false,
      description: "",
      key: "propertyName",
      label: "Property Name",
      type: "text",
    },
  ],
};

describe("PreviewConfig --- Snapshot", () => {
  it("Capturing Snapshot of Task Templates", async () => {
    const { baseElement, getByText } = rtlContextRouterRender(<PreviewConfig {...props} />);
    fireEvent.click(getByText("Preview"));
    expect(baseElement).toMatchSnapshot();
  });
});
