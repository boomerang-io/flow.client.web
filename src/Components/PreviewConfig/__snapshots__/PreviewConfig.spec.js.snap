// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewConfig --- Snapshot Capturing Snapshot of Task Templates 1`] = `
<body
  aria-hidden="true"
  class="bx--bmrg-body-modal-is-open"
  id="app"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div>
    <button
      class="bx--btn bx--btn--field bx--btn--ghost"
      style="width: 6.25rem;"
      tabindex="0"
      type="button"
    >
      Preview
      <svg
        aria-hidden="true"
        aria-label="Preview task"
        class="bx--btn__icon"
        fill="currentColor"
        focusable="false"
        height="16"
        preserveAspectRatio="xMidYMid meet"
        role="img"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15.5,7.8C14.3,4.7,11.3,2.6,8,2.5C4.7,2.6,1.7,4.7,0.5,7.8c0,0.1,0,0.2,0,0.3c1.2,3.1,4.1,5.2,7.5,5.3	c3.3-0.1,6.3-2.2,7.5-5.3C15.5,8.1,15.5,7.9,15.5,7.8z M8,12.5c-2.7,0-5.4-2-6.5-4.5c1-2.5,3.8-4.5,6.5-4.5s5.4,2,6.5,4.5	C13.4,10.5,10.6,12.5,8,12.5z"
        />
        <path
          d="M8,5C6.3,5,5,6.3,5,8s1.3,3,3,3s3-1.3,3-3S9.7,5,8,5z M8,10c-1.1,0-2-0.9-2-2s0.9-2,2-2s2,0.9,2,2S9.1,10,8,10z"
        />
      </svg>
    </button>
  </div>
  <div
    class="bx--bmrg-modal-portal"
  >
    <div
      class="bx--bmrg-modal-overlay --is-open"
    >
      <div
        aria-modal="true"
        class="bx--bmrg-modal-container bx--bmrg-modal-composed-container bx--modal-container modal-container-fix-width --is-open"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="bx--modal-header"
        >
          <p
            class="bx--modal-header__heading bx--type-beta"
          >
            [Preview] Test template
          </p>
          <p
            class="bx--bmrg-modal-composed-subtitle"
          >
            This is a preview of what the user sees when editing this Task. The user can also give this task a custom name for their Workflow, and can adjust its connected tasks. You can type in these fields to test any validation requirements.
          </p>
          <button
            aria-label="Close"
            class="bx--modal-close"
            title="Close"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="bx--modal-close__icon"
              fill="currentColor"
              focusable="false"
              height="20"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 32 32"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4 14.6 16 8 22.6 9.4 24 16 17.4 22.6 24 24 22.6 17.4 16 24 9.4z"
              />
            </svg>
          </button>
        </div>
        <div
          class="bx--bmrg-dynamic-formik"
        >
          <form
            class="bx--bmrg-modal-form"
            novalidate=""
          >
            <div
              class="bx--modal-content"
            >
              <div
                class="bx--bmrg-text-input"
              >
                <div
                  class="bx--form-item bx--text-input-wrapper"
                >
                  <label
                    class="bx--label"
                    for="['path']"
                  >
                    <div
                      class="bx--bmrg-text-input__label"
                    >
                      <div>
                        File Path
                      </div>
                      
                    </div>
                  </label>
                  <div
                    class="bx--text-input__field-outer-wrapper"
                  >
                    <div
                      class="bx--text-input__field-wrapper"
                    >
                      <input
                        class="bx--text-input bx--text__input"
                        id="['path']"
                        name="['path']"
                        placeholder=""
                        title=""
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="bx--bmrg-text-input"
              >
                <div
                  class="bx--form-item bx--text-input-wrapper"
                >
                  <label
                    class="bx--label"
                    for="['propertyName']"
                  >
                    <div
                      class="bx--bmrg-text-input__label"
                    >
                      <div>
                        Property Name
                      </div>
                      
                    </div>
                  </label>
                  <div
                    class="bx--text-input__field-outer-wrapper"
                  >
                    <div
                      class="bx--text-input__field-wrapper"
                    >
                      <input
                        class="bx--text-input bx--text__input"
                        id="['propertyName']"
                        name="['propertyName']"
                        placeholder=""
                        title=""
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    class="bx--bmrg-modal-portal"
  />
</body>
`;
